### PR TITLE
Enable jupyter kernel

### LINKED ISSUE #5: Add cadabra2-jupyter-kernel

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Fixes https://github.com/conda-forge/cadabra2-feedstock/issues/5
Fixes #8

### CODE REVIEW

**Comment by kpeeters** [comment]
I think the `cadabra-jupyter-kernel` needs `xeus` as runtime dependency. Moreover, I don't think you need to include `pybind11` in the build deps, as it is included in the cadabra source tree.

**Comment by isuruf** [comment]
> I think the cadabra-jupyter-kernel needs xeus as runtime dependency.

`xeus` is added using `run_exports`. i.e. if it is in `host`, it is added to `run` automatically. (You might have used an older `xeus` conda package without `run_exports`)

> Moreover, I don't think you need to include pybind11 in the build deps, as it is included in the cadabra source tree.

Thanks

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by isuruf** [comment]
@kpeeters, is that the only dependency vendored in cadabra2?

**Comment by kpeeters** [comment]
@isuruf There are a few others bundled as well (jsoncpp, sqlite3) but those are best not used for conda (those are essentially only there to make it easier to build on windows, where getting the right version of these libraries turned out to be a hassle).

**Comment by isuruf** [comment]
How do I make the build script not use the bundled ones? I want to use as much conda packages as possible.

**Comment by isuruf** [comment]
Btw, can you ship pybind11's license file as well? Due to pybind11 being BSD-3-Clause, that file needs to be shipped in the cadabra2 conda package.

**Comment by kpeeters** [comment]
You already do. For `pybind11` I would really like to stick to the bundled version as cadabra does a lot of very tricky python/c++ bridging things and it's best that we can rely on a particular well-tested version (and it makes no difference for the final result anyway as it is a header-only library, so it all gets included in the binary). The same holds true for `websocketpp` (which I didn't mention yet). The things that you can link to dynamically are already doing that with your current cmake options.

**Comment by kpeeters** [comment]
Thanks, totally missed that one, now in `libs/pybind11/LICENSE`.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -54,35 +53,134 @@ requirements:
...
 
 about:
   home: https://github.com/kpeeters/cadabra2
   license: GPL-3.0-only
   license_family: GPL
-  license_file: LICENSE
+  license_file:
+    - LICENSE
+    - libs/websocketpp/COPYING
+    - libs/pybind11/LICENSE
```
Comment (isuruf) [inline recipe/meta.yaml:184]: @kpeeters, is this correct? Only these 3?

Comment (kpeeters) [reply recipe/meta.yaml:184]: I have added `libs/jsoncpp/LICENSE`, `libs/linenoise/LICENSE`, `libs/tiny-process-library/LICENSE` and `libs/whereami/LICENSE.WTFPLv2`. That should cover it.

Comment (isuruf) [reply recipe/meta.yaml:184]: `libs/jsoncpp/LICENSE` is not needed because we are using the conda package here, right?

Comment (kpeeters) [reply recipe/meta.yaml:184]: Yes, I just added it to be complete.

**Comment by isuruf** [comment]
This is pending a new cadabra2 release now.

**Comment by kpeeters** [comment]
Ok, will do that tomorrow.

File: recipe/meta.yaml (hunk 1 above)
Comment (isuruf) [reply recipe/meta.yaml:184]: `libs/linenoise/LICENSE` is missing in upstream repo

**Comment by kpeeters** [comment]
I have released 2.3.0.

**Comment by isuruf** [comment]
Thanks. The files you mentioned at https://github.com/conda-forge/cadabra2-feedstock/pull/7#discussion_r436351089 are not in the repo, though.

**Comment by kpeeters** [comment]
I have no idea how it managed to skip that commit, but it's there now. Tagged commit is unchanged. I have uploaded the master tarball as release asset to give you a stable SHA.

**Comment by isuruf** [comment]
One last question. What do you think of the names `cadabra2` (for the python module), `cadabra2-gtk`, `cadabra-jupyter-kernel`?

**Comment by kpeeters** [comment]
You may want to change that to `cadabra2-jupyter-kernel`, just in case people think this is a kernel for the 1.x series of cadabra (which I am not maintaining anymore).

**Comment by kpeeters** [comment]
By the way, the `cadabra2` package also includes the command line client `cadabra2`, so it's not just the python module.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -54,35 +54,137 @@ requirements:
...
+      run_constrained:
+        - cadabra-jupyter-kernel {{ version }}
+    files:
+      - bin/cadabra2-gtk
+      - bin/cadabra2latex
+      - bin/cadabra-server
+      - bin/cadabra2cadabra
+      - bin/cadabra2html
+      - bin/cadabra2python
+      - bin/cadabra2-cli
```
Comment (isuruf) [inline recipe/meta.yaml:133]: @kpeeters, do you mean that this executable should be in `cadabra2` package?

Comment (isuruf) [reply recipe/meta.yaml:133]: Ah, never mind, `bin/cadabra2` is in `cadabra2` package.

**Comment by kpeeters** [comment]
Once this is visible for the public, it may be good to remove the existing `cadabra2` package from conda-forge. Or does that get overwritten automatically?

**Comment by isuruf** [comment]
`cadabra2-2.3.0` will be chosen instead of `cadabra2-2.2.8`

**Comment by kpeeters** [comment]
Ok. I have removed my `cadabra2-jupyter-kernel` package in channel `kpeeters`. Thanks for your help with all of this, much appreciated!

We should talk symengine at some point, I have some ideas (not conda-related).

**Comment by isuruf** [comment]
> We should talk symengine at some point, I have some ideas (not conda-related).

Yes, I'll be happy to have a chat.

**Comment by StefanoDeAngelis** [comment]
Hi! I tried to do the installation through Conda, but once I set the channel priority to ’strict’ the installation fails in ’solving the environment’. I experienced this problem both on Ubuntu and macOS.

On the other hand, with the ‘flexible’ channel priority, I can install ‘cadabra2', but I find the same problem when I am installing ‘cadabra2-gtk’.

Could you help me with this?

**Comment by isuruf** [comment]
Can you try to create a new environment instead of installing `cadabra2-gtk` to an existing environment?

**Comment by StefanoDeAngelis** [comment]
I created a new environment, but to make the installation work I had to set the channel priority to flexible. Then it worked.

**Comment by isuruf** [comment]
@StefanoDeAngelis, can you post the conda list after the insallation?

**Comment by isuruf** [comment]
In a new issue with all the details required in the issue template.

**Comment by isuruf** [comment]
@StefanoDeAngelis, ping on this

**Comment by StefanoDeAngelis** [comment]
@isuruf Sorry for the late reply. I created a new issue with all the info in 'Conda list' and 'Conda info', as you asked.

**Comment by isuruf** [comment]
Sorry. I wasn't watching this repo. I am now.

**Pull request merged**